astronomy/orbit_analysis_test.cpp

```diff
@@ -0,0 +1,509 @@
+        ephemeris_(earth_1957_.MakeEphemeris(
+            /*accuracy_parameters=*/{/*fitting_tolerance=*/1 * Milli(Metre),
+                                     /*geopotential_tolerance=*/0x1p-24},
+            Ephemeris<ICRS>::FixedStepParameters(
+                SymmetricLinearMultistepIntegrator<QuinlanTremaine1990Order12,
+                                                   Position<ICRS>>(),
+                /*step=*/1 * JulianYear))),
+        earth_(*earth_1957_.rotating_body(*ephemeris_, "Earth")) {}
+
+  // Returns a GCRS trajectory obtained by stitching together the trajectories
+  // of |sp3_orbit.satellites| in |sp3_orbit.files|.
+  not_null<std::unique_ptr<DiscreteTrajectory<GCRS>>> EarthCentredTrajectory(
+      SP3Orbit const& sp3_orbit) {
+    BodyCentredNonRotatingDynamicFrame<ICRS, GCRS> gcrs{ephemeris_.get(),
+                                                        &earth_};
+    BodySurfaceDynamicFrame<ICRS, ITRS> itrs{ephemeris_.get(), &earth_};
+
+    auto result = make_not_null_unique<DiscreteTrajectory<GCRS>>();
+    for (auto const& file : sp3_orbit.files.names) {
+      StandardProduct3 sp3(
+          SOLUTION_DIR / "astronomy" / "standard_product_3" / file,
+          sp3_orbit.files.dialect);
+      auto const& orbit = sp3.orbit(sp3_orbit.satellite);
+      CHECK_EQ(orbit.size(), 1);
+      auto const& arc = *orbit.front();
+      for (auto it = arc.Begin(); it != arc.End(); ++it) {
+        ephemeris_->Prolong(it.time());
+        result->Append(
+            it.time(),
+            gcrs.ToThisFrameAtTime(it.time())(
+                itrs.FromThisFrameAtTime(it.time())(it.degrees_of_freedom())));
+      }
+    }
+    return result;
+  }
+
+  std::pair<OrbitalElements, OrbitRecurrence> ElementsAndRecurrence(
+      SP3Orbit const& orbit) {
+    auto const elements =
+        OrbitalElements::ForTrajectory(*EarthCentredTrajectory(orbit),
+                                       earth_,
+                                       MasslessBody{}).ValueOrDie();
+    auto const recurrence =
+        OrbitRecurrence::ClosestRecurrence(elements.nodal_period(),
+                                           elements.nodal_precession(),
+                                           earth_,
+                                           /*max_abs_Cᴛₒ=*/100);
+    return {elements, recurrence};
+  }
+
+  SolarSystem<ICRS> earth_1957_;
+  not_null<std::unique_ptr<Ephemeris<ICRS>>> ephemeris_;
+  RotatingBody<ICRS> const& earth_;
+
+ private:
+  SolarSystem<ICRS> RemoveAllButEarth(SolarSystem<ICRS> solar_system) {
+    std::vector<std::string> const names = solar_system.names();
+    for (auto const& name : names) {
+      if (name != "Earth") {
+        solar_system.RemoveMassiveBody(name);
+      }
+    }
+    return solar_system;
+  }
+};
+
+// COSPAR ID 2010-001A, SVN C003.
+// 北斗二號 GEO01.
+// PRN C01, GEO, 140.0° E.
+TEST_F(OrbitAnalysisTest, 北斗GEO) {
+  auto [elements, recurrence] = ElementsAndRecurrence(  // NOLINT
```

Comment:
Should we remove the lint check that misfires here?

Reply:
I think so, or at least we should patch the linter.

It's `whitespace/braces`.

Reply:
Opened #2287.